Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp

```diff
@@ -0,0 +1,419 @@
+        // TODO
+        break;
+    case AST::ParameterizedType::Base::Mat4x3:
+        // TODO
+        break;
+    case AST::ParameterizedType::Base::Mat4x4:
+        // TODO
+        break;
+    }
+}
+
+void FunctionDefinitionWriter::visit(AST::Parameter& parameter)
+{
+    visit(parameter.type());
+    m_stringBuilder.append(" ", parameter.name());
+    for (auto& attribute : parameter.attributes()) {
+        m_stringBuilder.append(" ");
+        checkErrorAndVisit(attribute);
+    }
+}
+
+void FunctionDefinitionWriter::visit(AST::Expression& expression)
+{
+    AST::Visitor::visit(expression);
+}
+
+void FunctionDefinitionWriter::visit(AST::CallableExpression& call)
+{
+    bool first = true;
+    if (call.target().kind() == AST::Node::Kind::ArrayType) {
+        m_stringBuilder.append("{\n");
+        {
+            IndentationScope scope(m_indent);
+            for (auto& argument : call.arguments()) {
+                m_stringBuilder.append(m_indent);
+                visit(argument);
+                m_stringBuilder.append(",\n");
+                first = false;
+            }
+        }
+        m_stringBuilder.append(m_indent, "}");
+    } else {
+        visit(call.target());
+        m_stringBuilder.append("(");
+        for (auto& argument : call.arguments()) {
+            if (!first)
+                m_stringBuilder.append(", ");
+            visit(argument);
+            first = false;
+        }
+        m_stringBuilder.append(")");
+    }
+}
+
+void FunctionDefinitionWriter::visit(AST::UnaryExpression& unary)
+{
+    switch (unary.operation()) {
+    case AST::UnaryOperation::Negate:
+        m_stringBuilder.append("-");
+    }
+    visit(unary.expression());
+}
+
+void FunctionDefinitionWriter::visit(AST::ArrayAccess& access)
+{
+    visit(access.base());
+    m_stringBuilder.append("[");
+    visit(access.index());
+    m_stringBuilder.append("]");
+}
+
+void FunctionDefinitionWriter::visit(AST::IdentifierExpression& identifier)
+{
+    m_stringBuilder.append(identifier.identifier());
+}
+
+void FunctionDefinitionWriter::visit(AST::StructureAccess& access)
+{
+    visit(access.base());
+    m_stringBuilder.append(".", access.fieldName());
+}
+
+void FunctionDefinitionWriter::visit(AST::AbstractIntLiteral& literal)
+{
+    m_stringBuilder.append(literal.value());
+}
+
+void FunctionDefinitionWriter::visit(AST::Int32Literal& literal)
+{
+    m_stringBuilder.append(literal.value());
+}
+
+void FunctionDefinitionWriter::visit(AST::AbstractFloatLiteral& literal)
+{
+    m_stringBuilder.append(literal.value());
+}
+
+void FunctionDefinitionWriter::visit(AST::Float32Literal& literal)
+{
+    m_stringBuilder.append(literal.value());
+}
+
+void FunctionDefinitionWriter::visit(AST::Statement& statement)
+{
+    if (statement.kind() != AST::Node::Kind::CompoundStatement)
+        m_stringBuilder.append(m_indent);
+
+    AST::Visitor::visit(statement);
+
+    if (statement.kind() != AST::Node::Kind::CompoundStatement)
```

Comment:
What is an example of a compound statement?

Reply:
A compound statement is a list of statements, e.g. the body of a function